Docker image build hangs and fails on go test

**Describe the bug**
When running `make images`, the `onos-config` Docker build consistently fails for me on the master branch. It fails at different points during `go test` (examples below).

**To Reproduce**
Steps to reproduce the behavior:
1. Pull `master`
2. Run `make images`
3. Repeat if necessary
4. See error

**Expected behavior**
No error :-)

**Screenshots**

```
go test github.com/onosproject/onos-config/pkg/...
?   	github.com/onosproject/onos-config/pkg	[no test files]
The command '/bin/sh -c cd /go/src/github.com/onosproject/onos-config && make build' returned a non-zero code: 137
make: *** [onos-config-docker] Error 137
```

```
go test github.com/onosproject/onos-config/pkg/...
?   	github.com/onosproject/onos-config/pkg	[no test files]
?   	github.com/onosproject/onos-config/pkg/certs	[no test files]
?   	github.com/onosproject/onos-config/pkg/cli	[no test files]
?   	github.com/onosproject/onos-config/pkg/cli/command	[no test files]
ok  	github.com/onosproject/onos-config/pkg/dispatcher	0.018s
ok  	github.com/onosproject/onos-config/pkg/events	0.238s
ok  	github.com/onosproject/onos-config/pkg/manager	0.119s
?   	github.com/onosproject/onos-config/pkg/northbound	[no test files]
The command '/bin/sh -c cd /go/src/github.com/onosproject/onos-config && make build' returned a non-zero code: 137
make: *** [onos-config-docker] Error 137
```

--- a/cmd/onos-config/onos-config.go
+++ b/cmd/onos-config/onos-config.go
@@ -127,5 +127,7 @@ func startServer(caPath string, keyPath string, certPath string) error {
 	s.AddService(diags.Service{})
 	s.AddService(gnmi.Service{})
 
-	return s.Serve()
+	return s.Serve(func(started string) {
+		log.Info("Started NBI on", started)
+	})
 }

--- a/pkg/northbound/admin/admin_test.go
+++ b/pkg/northbound/admin/admin_test.go
@@ -22,12 +22,16 @@ import (
 	"gotest.tools/assert"
 	"io"
 	"os"
+	"sync"
 	"testing"
 )
 
 // TestMain initializes the test suite context.
 func TestMain(m *testing.M) {
-	northbound.SetUpServer(10124, Service{})
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(1)
+	northbound.SetUpServer(10124, Service{}, &waitGroup)
+	waitGroup.Wait()
 	os.Exit(m.Run())
 }
 
@@ -45,33 +49,33 @@ func Test_GetNetworkChanges(t *testing.T) {
 	conn, client := getAdminClient()
 	defer conn.Close()
 	stream, err := client.GetNetworkChanges(context.Background(), &proto.NetworkChangesRequest{})
-	assert.Assert(t, err == nil && stream != nil, "unable to issue request")
-	var name = ""
+	assert.NilError(t, err, "unable to issue request")
+	var name string
 	for {
 		in, err := stream.Recv()
 		if err == io.EOF || in == nil {
 			break
 		}
-		assert.Assert(t, err == nil, "unable to receive message")
+		assert.NilError(t, err, "unable to receive message")
 		name = in.Name
 	}
 	err = stream.CloseSend()
-	assert.Assert(t, err == nil, "unable to close stream")
-	assert.Assert(t, len(name) > 0, "wrong name received")
+	assert.NilError(t, err, "unable to close stream")
+	assert.Assert(t, len(name) > 0, "no name received")
 }
 
 func Test_RollbackNetworkChange_BadName(t *testing.T) {
 	conn, client := getAdminClient()
 	defer conn.Close()
 	_, err := client.RollbackNetworkChange(context.Background(), &proto.RollbackRequest{Name: "BAD CHANGE"})
-	assert.Assert(t, err != nil, "should not rollback")
+	assert.ErrorContains(t, err, "Rollback aborted. Network change BAD CHANGE not found")
 }
 
 func Test_RollbackNetworkChange_NoChange(t *testing.T) {
 	conn, client := getAdminClient()
 	defer conn.Close()
 	_, err := client.RollbackNetworkChange(context.Background(), &proto.RollbackRequest{Name: ""})
-	assert.Assert(t, err != nil, "should not rollback")
+	assert.ErrorContains(t, err, "is not")
 }
 
 func Test_AddDevice(t *testing.T) {
@@ -81,28 +85,28 @@ func Test_AddDevice(t *testing.T) {
 	oldCount := resp.Count
 	_, err = client.AddOrUpdateDevice(context.Background(),
 		&proto.DeviceInfo{Id: "device", Address: "address", Version: "0.9"})
-	assert.Assert(t, err == nil, "should add device")
+	assert.NilError(t, err, "should add device")
 	resp, err = client.GetDeviceSummary(context.Background(), &proto.DeviceSummaryRequest{})
-	assert.Assert(t, oldCount+1 == resp.Count, "should add new device")
+	assert.Equal(t, oldCount+1, resp.Count, "should add new device")
 }
 
 func Test_GetDevices(t *testing.T) {
 	conn, client := getDeviceClient()
 	defer conn.Close()
 	stream, err := client.GetDevices(context.Background(), &proto.GetDevicesRequest{})
-	assert.Assert(t, err == nil && stream != nil, "unable to issue request")
-	var id = ""
+	assert.NilError(t, err, "unable to issue request")
+	var id string
 	for {
 		in, err := stream.Recv()
 		if err == io.EOF || in == nil {
 			break
 		}
-		assert.Assert(t, err == nil, "unable to receive message")
+		assert.NilError(t, err, "unable to receive message")
 		id = in.Id
 	}
 	err = stream.CloseSend()
-	assert.Assert(t, err == nil, "unable to close stream")
-	assert.Assert(t, len(id) > 0, "wrong id received")
+	assert.NilError(t, err, "unable to close stream")
+	assert.Assert(t, len(id) > 0, "no id received")
 }
 
 func Test_RemoveDevice(t *testing.T) {
@@ -110,11 +114,11 @@ func Test_RemoveDevice(t *testing.T) {
 	defer conn.Close()
 	_, err := client.AddOrUpdateDevice(context.Background(),
 		&proto.DeviceInfo{Id: "device", Address: "address", Version: "0.9"})
-	assert.Assert(t, err == nil, "should add device")
+	assert.NilError(t, err, "should add device")
 	resp, err := client.GetDeviceSummary(context.Background(), &proto.DeviceSummaryRequest{})
 	oldCount := resp.Count
 	_, err = client.RemoveDevice(context.Background(), &proto.DeviceInfo{Id: "device"})
-	assert.Assert(t, err == nil, "should remove device")
+	assert.NilError(t, err, "should remove device")
 	resp, err = client.GetDeviceSummary(context.Background(), &proto.DeviceSummaryRequest{})
-	assert.Assert(t, oldCount-1 == resp.Count, "should remove existing device")
+	assert.Equal(t, oldCount-1, resp.Count, "should remove existing device")
 }

--- a/pkg/northbound/diags/diags_test.go
+++ b/pkg/northbound/diags/diags_test.go
@@ -22,12 +22,16 @@ import (
 	"gotest.tools/assert"
 	"io"
 	"os"
+	"sync"
 	"testing"
 )
 
 // TestMain initializes the test suite context.
 func TestMain(m *testing.M) {
-	northbound.SetUpServer(10123, Service{})
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(1)
+	northbound.SetUpServer(10123, Service{}, &waitGroup)
+	waitGroup.Wait()
 	os.Exit(m.Run())
 }
 
@@ -54,19 +58,19 @@ func testGetChanges(t *testing.T, changeID string) {
 	}
 
 	stream, err := client.GetChanges(context.Background(), changesReq)
-	assert.Assert(t, err == nil && stream != nil, "unable to issue request")
+	assert.NilError(t, err, "unable to issue request")
 	var id = ""
 	for {
 		in, err := stream.Recv()
 		if err == io.EOF || in == nil {
 			break
 		}
-		assert.Assert(t, err == nil, "unable to receive message")
+		assert.NilError(t, err, "unable to receive message")
 		id = in.Id
 	}
 	err = stream.CloseSend()
-	assert.Assert(t, err == nil, "unable to close stream")
-	assert.Assert(t, len(id) > 0, "wrong id received")
+	assert.NilError(t, err, "unable to close stream")
+	assert.Assert(t, len(id) > 0, "no id received")
 }
 
 func Test_GetConfigurations_All(t *testing.T) {
@@ -86,17 +90,17 @@ func testGetConfigurations(t *testing.T, deviceID string) {
 		configReq.DeviceIds = append(configReq.DeviceIds, deviceID)
 	}
 	stream, err := client.GetConfigurations(context.Background(), configReq)
-	assert.Assert(t, err == nil && stream != nil, "unable to issue request")
+	assert.NilError(t, err, "unable to issue request")
 	var name = ""
 	for {
 		in, err := stream.Recv()
 		if err == io.EOF || in == nil {
 			break
 		}
-		assert.Assert(t, err == nil, "unable to receive message")
+		assert.NilError(t, err, "unable to receive message")
 		name = in.Name
 	}
 	err = stream.CloseSend()
-	assert.Assert(t, err == nil, "unable to close stream")
-	assert.Assert(t, len(name) > 0, "wrong name received")
+	assert.NilError(t, err, "unable to close stream")
+	assert.Assert(t, len(name) > 0, "no name received")
 }

--- a/pkg/northbound/server.go
+++ b/pkg/northbound/server.go
@@ -77,7 +77,7 @@ func (s *Server) AddService(r Service) {
 }
 
 // Serve starts the NB gNMI server.
-func (s *Server) Serve() error {
+func (s *Server) Serve(started func(string)) error {
 	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", s.cfg.Port))
 	if err != nil {
 		return err
@@ -123,6 +123,7 @@ func (s *Server) Serve() error {
 	for i := range s.services {
 		s.services[i].Register(server)
 	}
+	started(lis.Addr().String())
 
 	log.Infof("Starting RPC server on address: %s", lis.Addr().String())
 	return server.Serve(lis)

--- a/pkg/northbound/testUtils.go
+++ b/pkg/northbound/testUtils.go
@@ -20,7 +20,7 @@ import (
 	"github.com/onosproject/onos-config/pkg/manager"
 	"google.golang.org/grpc"
 	"log"
-	"time"
+	"sync"
 )
 
 var (
@@ -33,7 +33,7 @@ var (
 
 // SetUpServer sets up a test manager and a gRPC end-point
 // to which it registers the given service.
-func SetUpServer(port int16, service Service) {
+func SetUpServer(port int16, service Service, waitGroup *sync.WaitGroup) {
 	var err error
 	_, err = manager.LoadManager(
 		"../../../configs/configStore-sample.json",
@@ -56,12 +56,13 @@ func SetUpServer(port int16, service Service) {
 	if err != nil {
 		log.Fatal("Error loading cert", err)
 	}
-
 	go func() {
-		err := s.Serve()
+		err := s.Serve(func(started string) {
+			waitGroup.Done()
+			fmt.Printf("Started %v", started)
+		})
 		if err != nil {
 			log.Fatal("Unable to serve", err)
 		}
 	}()
-	time.Sleep(5000000) // FIXME: this is clearly a hack, we need to replace this with wait on chan
 }

